File: Final Exam/p3.py
#' #Problem 3 Take Home Final Exam

# Importing the necessary libraries
import matplotlib.pyplot as plt
import numpy as np
from numpy.random import randn, uniform, seed
from numpy.linalg import norm, solve, matrix_rank, inv

#' For this problem we need to solve:
#' 		min f(x) = sum(xi log(xi)) over i=1,...,n
#'		subject to Ax=b
#' 		A in R pxn
#'		p<n

#' Define the new functions
def f(x):
	return np.dot(np.transpose(x),np.log(x)) # This dot product is equivalent to summation over i because x is length n

def grad_f(x):
	return np.log(x) + np.ones((n,1))

def hess_f(x):
	return np.diag((1/x)[:,0])

#' Initialize the problem
n = 100
p = 30

#' Choose A randomly
seed(999)
A = randn(p,n)

#' Make sure that A is full rank
while matrix_rank(A)<p and matrix_rank(A)<n:
	A = randn(p,n)

#' Set the range for the distribution on random x
low = 0.0
high = 1.0
x = uniform(low,high,size=(n,1))

#' Set b=Axhat
b = np.dot(A,x)

#' ## Part A - Using the standard newton's method
# Backtracking method for updating t
def backtrack(x,f,grad_f,dx,alpha,beta):
	t=1.0
	y = f(x)
	gx = grad_f(x)
	while f(x+t*dx) > y+alpha*t*np.dot(np.transpose(gx),dx):
		t=beta*t
	return t

# Newton Method
# This time the function only returns the optimal point found
def newton_method(x,iterations,alpha,beta,eps):
	# Repeat
	y = np.array([])
	s = np.array([])

	hx = inv(hess_f(x))
	gx = grad_f(x)

	dnt = -np.dot(hx,gx)
	dec = -np.dot(np.transpose(gx),dnt)

	p = f(x)
	y = np.append(y,p)

	for i in range(iterations):

		if dec/2 <= eps:
			break

		t = backtrack(x,f,grad_f,dnt,alpha,beta)
		s = np.append(s,t)

		x = x + t*dnt

		p = f(x)
		y = np.append(y,p)

		hx = inv(hess_f(x))
		gx = grad_f(x)

		dnt = -np.dot(hx,gx)
		dec = -np.dot(np.transpose(gx),dnt)

	return y,s,p

# Set the parameters for Newton's Method
max_i = 1000
alpha =0.1
beta = 0.25
eps = 1e-8

y,s,p_star = newton_method(x,max_i,alpha,beta,eps)

fig,ax = plt.subplots()
ax.semilogy(np.transpose(y-p_star))
ax.set_title('Convergence to Optimal Solution')
ax.set_xlabel('Iterations')
ax.set_ylabel('f - p_star')
plt.show()

#' I couldn't get the inseasible Newton method to work on here and debugging via python is terrible. I am going to implement b and c in matlab.

File: Final Exam/p3.m
%% Problem 3, this time in matlab

clear all; close all;

% Set parameters
n = 100;
p = 30;
alpha = 0.05;
beta = .5;
iters = 1000;
epsilon = 1e-8;

% Choose A randomly
rng('default')
A = randn(p,n);

% Check if A is full rank
while(rank(A)<p && rank(A)<n)
    A = randn(p,n);
end

% Get random estimate of x
x = rand(n,1); % Choose x in [0,1]

% Set b using the estimated x
b = A*x;

% Get function values
f = @(x) x'*log(x);
grad = @(x) log(x) + ones(length(x),1);
hess = @(x) diag(1./x);

%% Newton's method
y = [];
s = [];
d = [];
p_star = f(x);
y = [y; p_star];
h = inv(hess(x));
g = grad(x);

dnt = -h*g;
dec = -1*g'*dnt;
d = [d;dec];

for(i=1:iters)
    if(dec/2<=epsilon)
        break;
    end
    
    t = 1;
    
    while(f(x+t*dnt)>p_star+alpha*t*g'*dnt)
        t = beta*t;
    end
    
    x = x+t*dnt;
    
    p_star = f(x);
    g = grad(x);
    h = inv(hess(x));
    
    dnt = -h*g;
    dec = -1*g'*dnt;
    d = [d;dec];

    s = [s;t];
    y = [y;p_star];    
    
end


semilogy(y-p_star)
hold on

%% Now want to use the infeasible start newton method

% Compute the primal and dual netwon steps
nu = 10*ones(p,1);
y = [];
d1 = [];

% Compute residuals
r_d = @(x,nu) A'*nu + grad(x);
r_p = @(x) A*x-b;
R = @(x,nu) -1*[r_d(x,nu); r_p(x)];

% Create matrix to solve to deltas
Obj = @(hx) [hx,A';A,zeros(p,p)];
y = [y;f(x)];

for(i=1:iters)

    % Get primal and dual deltas
    hx = hess(x);
    O_dum = Obj(hx);
    R_dum = R(x,nu);
    D = O_dum\R_dum;
    dnt = D(1:n);
    nnt = D(n+1:n+p);

    % Backtracking
    t = 1;
    while(norm(R(x+t*dnt,nu+t*nnt))>(1-alpha*t)*norm(R(x,nu)) && min(x+t*dnt)<=0)
        t = beta*t;
    end

    x = x+dnt;
    nu = nu+nnt;
    
    p_star1 = f(x);
    y = [y;p_star1];
    
    if r_p(x)<=1e-8 & norm(R(x,nu))<=epsilon
        break;
    end
end


for(i=1:iters)
    if(dec/2<=epsilon)
        break;
    end
    
    t = 1;
    
    while(f(x+t*dnt)>p_star1+alpha*t*g'*dnt)
        t = beta*t;
    end
    
    x = x+t*dnt;
    
    p_star1 = f(x);
    g = grad(x);
    h = inv(hess(x));
    
    dnt = -h*g;
    dec = -1*g'*dnt;

    s = [s;t];
    y = [y;p_star1];    
    
end

semilogy(real(y-p_star1))
title('Plot of Newton Method and Infeasible Newton Method Convergence')